README
This is Perl module App::YesssGuthaben.

INSTALLATION

App::YesssGuthaben installation is straightforward. If your CPAN shell is set up,
you should just be able to do

    % cpan App::YesssGuthaben

Download it, unpack it, then build it as per the usual:

    % perl Build.PL
    % make && make test

Then install it:

    % sudo ./Build install

DOCUMENTATION

App::YesssGuthaben documentation is available as in POD. So you can do:

    % perldoc App::YesssGuthaben

to read the documentation online with your favorite pager.

Thomas Klausner


Changes
Revision history for Perl extension App::YesssGuthaben

0.02	2009-05-11 08:21:27
    - added info on SOLL (Thomas Klausner)


0.01    Tue May  5 15:28:03 2009
        - original version


lib/App/YesssGuthaben.pm
package App::YesssGuthaben;

use strict;
use warnings;
use 5.010;
use version; our $VERSION = version->new('0.01');
use WWW::Mechanize;
use Carp;
use DateTime;
use HTML::TableExtract;

use Moose;
with qw(
    MooseX::Getopt);

has msisdn => (is=>'ro',isa=>'Str',required=>1);
has passwd => (is=>'ro',isa=>'Str',required=>1);
has mech   => (is=>'rw',isa=>'WWW::Mechanize',default=>sub { WWW::Mechanize->new });
has yesss_login  => (is=>'ro',isa=>'Str',default=>'http://www.yesss.at/kontomanager.php');
has yesss_bookings=> (is=>'ro',isa=>'Str',default=>'https://www.yesss.at/kontomanager/wertkarte_gespraeche.php?page=');
has verbose => (is=>'rw',isa=>'Bool',default=>0);

no Moose;
__PACKAGE__->meta->make_immutable;

sub run {
    my $self = shift;

    $self->login;
    $self->get_this_months_bookings; 
}

sub login {
    my $self = shift;
    $self->mech->get($self->yesss_login);
    $self->mech->submit_form(
        with_fields=>{
            rufnummer=>$self->msisdn,
            passwort=>$self->passwd,
        }
    );
}

sub get_this_months_bookings {
    my $self = shift;

    my $last_day_of_prev_month = DateTime->now->truncate('to'=>'month')->subtract('days'=>1)->dmy('.');

    my $sum;
    PAGE: foreach my $page (1 .. 10) {
        my $table_extract = HTML::TableExtract->new(headers => ['Datum/Uhrzeit:',qw(Nummer: Dauer: Kosten: Art: Verrechnung:)]);
        $self->mech->get($self->yesss_bookings . $page);
        $table_extract->parse($self->mech->content);
        foreach my $row ($table_extract->rows) {
            my $date = substr($row->[0],0,10);
            last PAGE if $date eq $last_day_of_prev_month; 
            say join(';',$date,$row->[3],$row->[4]) if $self->verbose;
            $sum+=$row->[3];
        }
    }  
    say $sum;
}

1;

__END__

=head1 NAME

App::YesssGuthaben - Inform on this months yesss spendings

=head1 SYNOPSIS

  use App::YesssGuthaben;

=head1 DESCRIPTION

App::YesssGuthaben is a quick hack to screenscape the yesss.at 
(Austrian prepaid mobile provider) to get the sum of this months 
spendings (so I can tell my son when to stop texting!!!)

=head1 AUTHOR

Thomas Klausner E<lt>domm {at} cpan.orgE<gt>

=head1 SEE ALSO

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/App/YesssSpendings.pm
package App::YesssSpendings;

use strict;
use warnings;
use 5.010;
use version; our $VERSION = version->new('0.02');
use WWW::Mechanize;
use Carp;
use DateTime;
use HTML::TableExtract;

use Moose;
with qw(
    MooseX::Getopt);

has msisdn => (is=>'ro',isa=>'Str',required=>1);
has passwd => (is=>'ro',isa=>'Str',required=>1);
has mech   => (is=>'rw',isa=>'WWW::Mechanize',default=>sub { WWW::Mechanize->new });
has yesss_login  => (is=>'ro',isa=>'Str',default=>'http://www.yesss.at/kontomanager.php');
has yesss_bookings=> (is=>'ro',isa=>'Str',default=>'https://www.yesss.at/kontomanager/wertkarte_gespraeche.php?page=');
has verbose => (is=>'rw',isa=>'Bool',default=>0);
has budget => (is=>'ro',isa=>'Str');

no Moose;
__PACKAGE__->meta->make_immutable;

sub run {
    my $self = shift;

    $self->login;
    $self->get_this_months_bookings; 
}

sub login {
    my $self = shift;
    $self->mech->get($self->yesss_login);
    $self->mech->submit_form(
        with_fields=>{
            rufnummer=>$self->msisdn,
            passwort=>$self->passwd,
        }
    );
}

sub get_this_months_bookings {
    my $self = shift;

    my $last_day_of_prev_month = DateTime->now->truncate('to'=>'month')->subtract('days'=>1)->dmy('.');

    my $sum;
    PAGE: foreach my $page (1 .. 10) {
        my $table_extract = HTML::TableExtract->new(headers => ['Datum/Uhrzeit:',qw(Nummer: Dauer: Kosten: Art: Verrechnung:)]);
        $self->mech->get($self->yesss_bookings . $page);
        $table_extract->parse($self->mech->content);
        foreach my $row ($table_extract->rows) {
            my $date = substr($row->[0],0,10);
            last PAGE if $date eq $last_day_of_prev_month; 
            say join(';',$date,$row->[3],$row->[4]) if $self->verbose;
            $sum+=$row->[3];
        }
    }  
    
    # compare with budget
    if (my $budget = $self->budget) {
        my $now = DateTime->now->truncate('to'=>'day');
        my $days = DateTime->last_day_of_month(year=>$now->year,month=>$now->month)->day || 30;
        my $soll = sprintf("%5.3f",($budget/$days) * $now->day);
        say "$sum / $soll";
        my $rest = $soll - $sum;
        if ($rest > 0) {
            my $sms_left = int($rest / 0.068);
            say "Noch $sms_left SMS möglich";
        }
        else {
            say "STOP! Du bis drüber!";
        }
    }
    else {
        say $sum;
    }
}

1;

__END__

=head1 NAME

App::YesssSpendings - Inform on this months yesss spendings

=head1 SYNOPSIS

  use App::YesssSpendings;

=head1 DESCRIPTION

App::YesssSpendings is a quick hack to screenscape the yesss.at 
(Austrian prepaid mobile provider) to get the sum of this months 
spendings (so I can tell my son when to stop texting!!!)

=head1 AUTHOR

Thomas Klausner E<lt>domm {at} cpan.orgE<gt>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
